packages/vscode-extension/src/debug/localTelemetryReporter.ts

```diff
@@ -67,6 +91,145 @@ export async function sendDebugAllStartEvent(additionalProperties: {
+    const manifestString = manifestFile.getData().toString("utf-8");
+    return JSON.parse(manifestString);
+  } else {
+    return undefined;
+  }
+}
+
+async function readManifest(projectPath: string, manifestPath: string): Promise<TeamsAppManifest> {
+  if (!path.isAbsolute(manifestPath)) {
+    manifestPath = path.join(projectPath, manifestPath);
+  }
+  return await fs.readJson(manifestPath, { encoding: "utf-8" });
+}
+
+const defaultManifestPathsV3 = ["appPackage/manifest.json"];
+
+export const ManifestSources = Object.freeze({
+  ConfigureAppPackageManifestPath: "ConfigureAppPackageManifestPath",
+  ConfigureAppPackageAppPackagePath: "ConfigureAppPackageAppPackagePath",
+  DefaultManifestPath: "DefaultManifestPath",
+});
+export type ManifestSource = typeof ManifestSources[keyof typeof ManifestSources];
+
+// Find manifest by search in the yaml file with best effort.
+async function tryGetManifestFromYml(
+  projectPath: string,
+  yml: ProjectModel
+): Promise<{ source: ManifestSource; manifest: TeamsAppManifest } | undefined> {
+  const configureTeamsApp = yml.provision?.driverDefs?.find(
+    (item) => item.uses === updateAppPackageActionName
+  );
+  const configureTeamsAppArgs = configureTeamsApp?.with as
+    | Partial<ConfigureTeamsAppArgs>
+    | undefined;
+  const configureTeamsAppPath = configureTeamsAppArgs?.appPackagePath;
+
+  if (configureTeamsAppPath) {
+    // Case 1: Happy path
+    // Start from "teamsApp/update".appPackagePath
+    // => "teamsApp/zipAppPackage".outputZipPath
+    // => "teamsApp/zipAppPackage".manifestPath
+    try {
+      let manifestPath: string | undefined;
+      yml.provision?.driverDefs?.forEach((item) => {
+        if (item.uses !== createAppPackageActionName) {
+          return;
+        }
+        const createAppPackageArgs = item.with as Partial<CreateAppPackageArgs> | undefined;
+        if (!createAppPackageArgs?.outputZipPath) {
+          return;
+        }
+
+        if (createAppPackageArgs.outputZipPath === createAppPackageActionName) {
+          manifestPath = createAppPackageArgs.manifestPath;
+        }
+      });
+      if (manifestPath) {
+        return {
+          source: ManifestSources.ConfigureAppPackageManifestPath,
+          manifest: await readManifest(projectPath, manifestPath),
+        };
+      }
+    } catch (e) {
+      // fall through next case
+    }
+
+    // Case 2: Assume user uploads app package manually
+    // Start from "teamsApp/zipAppPackage".appPackagePath
+    // => Unzip appPackage (in memory) to get manifest
+    try {
+      const manifest = await readManifestFromAppPackage(configureTeamsAppPath);
+      if (manifest) {
+        return {
+          source: ManifestSources.ConfigureAppPackageAppPackagePath,
+          manifest,
+        };
+      }
+    } catch (e) {
+      // fall through next case
+    }
+  }
+
+  // Case 3: Assume user zips & uploads app package manually
+  // Try default location of manifest: appPackage/manifest.json
```

Comment:
Do you mean the template path or the build output path? E.g., `build/appPackage/manifest.xxx.json`?

Reply:
Template path. Because in this case, probably appPackage.zip isn't generated by the toolkit (user doesn't use "teamsApp/zipAppPackage"/"teamsApp/update").

Although template file may contain placeholder, it doesn't matter because we only use it to determine project type.

I will add a property later to indicate which is the method we get the manifest so they won't mix with normal cases and contaminate the data.
![image](https://user-images.githubusercontent.com/9698542/227081650-3bc9b1b7-a416-4681-bc0d-ae5978b59c54.png)